### PR TITLE
Loader modularization

### PR DESCRIPTION
**Pull request by dylandreimerink**
This PR turns the eBPF loader in `pkg/datapath/loader` into a Hive cell. It also unexported all symbols where possible. Global functions which were called outside of the package have been moved to the Loader interface so we have a single clear API boundary, making it easier to improve upon in the future.

This step will also make it easier to move non-loader logic outside of the package by allowing us to then still depend on the logic via hive.

Part of #30024

### CODE REVIEW

**Review by rgo3 — APPROVED** [review]
LGTM, pulled in Timo as well to have more loader :eyes: on this change.

**Review by asauber — APPROVED** [review]
cilium-dbg related changes lgtm

File: pkg/datapath/loader/netlink.go [hunk 1]
```diff
@@ -391,7 +391,7 @@ func setupVethPair(name, peerName string) error {
 // the first step of datapath initialization, then performs the setup (and
 // creation, if needed) of those interfaces. It returns two links and an error.
 // By default, it sets up the veth pair - cilium_host and cilium_net.
-func SetupBaseDevice(mtu int) (netlink.Link, netlink.Link, error) {
+func (l *loader) SetupBaseDevice(mtu int) (netlink.Link, netlink.Link, error) {
```
Comment (ti-mo) [inline pkg/datapath/loader/netlink.go]: We can avoid leaving this exported and adding this to the loader API at all by making `createDevices()` in `daemon_privileged_tests.go` create a dummy device named `defaults.HostDevice`. All other uses are internal, and for the sake of the privileged tests, any device will do. cilium_net isn't even touched.

File: pkg/datapath/loader/template.go [hunk 2]
```diff
@@ -282,6 +282,6 @@ func elfVariableSubstitutions(ep datapath.Endpoint) map[string]uint64 {
 // ELFSubstitutions fetches the set of variable and map substitutions that
 // must be implemented against an ELF template to configure the datapath for
 // the specified endpoint.
-func ELFSubstitutions(ep datapath.Endpoint) (map[string]uint64, map[string]string) {
+func (l *loader) ELFSubstitutions(ep datapath.Endpoint) (map[string]uint64, map[string]string) {
```
Comment (ti-mo) [inline pkg/datapath/loader/template.go:285]: Nit: can this and its dependent functions be moved to package endpoint?

**Review by ti-mo — APPROVED** [review]
Thanks! Left a few suggestions, but these can be picked up in follow-ups as well.

File: pkg/datapath/loader/template.go (hunk 2 above)
Comment (dylandreimerink) [reply pkg/datapath/loader/template.go:285]: Actually, its only used internally, except for in the tests of `pkg/datapath/linux/config`. And I don't see a what to not brick the tests. We should merge the header file writing logic of `pkg/datapath/linux/config` into `pkg/datapath/loader` at which point we can make it an internal function

**Review by markpash — APPROVED** [review]

**Review by mhofstetter — APPROVED** [review]
`sig-agent` :heavy_check_mark: 

Nice - thanks Dylan!

**Review by aanm — APPROVED** [review]

**Pull request merged**